source/state_representation/src/trajectory/CartesianTrajectory.cpp

```diff
@@ -0,0 +1,210 @@
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = new_point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = new_point.get_name();
+  this->TrajectoryBase<CartesianTrajectoryPoint>::add_point(trajectory_point);
+}
+
+void CartesianTrajectory::add_points(
+    const std::vector<CartesianState>& new_points, const std::vector<std::chrono::nanoseconds>& durations
+) {
+  if (new_points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (new_points.size() != durations.size()) {
+    throw exceptions::IncompatibleSizeException("The size of the points and durations vectors are not equal");
+  } else if (std::ranges::any_of(new_points, [&](const auto& p) { return p.is_empty(); })) {
+    throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+  } else if (!std::ranges::all_of(new_points, [&](const auto& p) {
+               return p.get_reference_frame() == this->reference_frame_;
+             })) {
+    throw exceptions::IncompatibleReferenceFramesException("Incompatible reference frames within the points vector");
+  }
+  for (unsigned int i = 0; i < new_points.size(); ++i) {
+    add_point(new_points[i], durations[i]);
+  }
+}
+
+void CartesianTrajectory::insert_point(
+    const CartesianState& new_point, const std::chrono::nanoseconds& duration, unsigned int index
+) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (new_point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + new_point.get_reference_frame() + " and " + this->reference_frame_
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = new_point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = new_point.get_name();
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::insert_point(trajectory_point, index);
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::set_point(
+    const CartesianState& point, const std::chrono::nanoseconds& duration, unsigned int index
+) {
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + point.get_reference_frame() + " and " + this->reference_frame_
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = point.get_name();
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::set_point(trajectory_point, index);
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::set_points(
```

Comment:
There is too much duplication between `set_point` and `set_points`. Either `set_points` has a for loop that calls `set_point` or the only thing that `set_point` does is `this->set_points({point}, {duration});`. My preference is the first